### ISSUE TITLE
Container networking stops too early on cell draining (CF update)

### ISSUE THREAD

**Issue by stephanme** [cloudfoundry/cf-networking-release#76]
## Issue

On CF update (cell draining), container networking stops to work for drained app instances before they are stopped.

## Context

The issue was first observed in a Java application that uses a [Hazelcast](https://hazelcast.org/imdg/) cluster (a distributed hashmap).
On CF updates, I see cluster outages due to lost connectivity between cluster nodes (= CF app instances). I don't observe similar outages when updating my application using `cf7 restart --strategy rolling`.

Follow-up for discussion that started on [#networking slack](https://cloudfoundry.slack.com/archives/CFX13JK7B/p1585904886012300).
CF 12.39

## Steps to Reproduce

See README.md in attached c2ctest.zip. The zip contains a test application but also a log file of one test run.

The test application starts a simple web server and probes connectivity to itself via internal (container networking) and external (gorouter) routes every 5s.
App instances are identified via CF_INSTANCE_INTERNAL_IP. Logs record the connectivity, i.e. which app instance requested and which app instance responded.

Connectivity behavior is observed during a rolling app restart (`cf7 restart c2ctest --strategy rolling`) and during a recreate of the Diego cell on which the app instance was running (simulating a CF update).

## Expected result

No connectivity errors between the two app instance during updates, no matter if the update is a CF update or a rolling application update.
Since Bosh DNS might be slow with reflecting the app instances, I expect at least one successful communication between the old and new app instance using container networking (internal routes) while updating.

## Current result

During CF update (bosh recreate diego-cell)
- Connection failures on internal routes with beginning of cell draining
- No successfull communication between the old and new app instance using internal routes

No problems observed during rolling application restart.

## Possible Fix

no idea

## Additional Context

[c2ctest.zip](https://github.com/cloudfoundry/cf-networking-release/files/4458569/c2ctest.zip) with an example app that demonstrates the issue.

**Comment by ameowlia** [comment]
Pasting the reproduction steps from the zip file here.


# Test program to check Container Networking

The application contains a simple web server and probes connectivity to itself via internal (container networking) and external (gorouter) routes every 5s.
App instances are identified via CF_INSTANCE_INTERNAL_IP. Logs record the connectivity, i.e. which app instance requested and which app instance responded.

Connectivity behavior is observed during a rolling app update (`cf7 restart c2ctest --strategy rolling`) and during a recreate of the Diego cell on which the app instance was running (simulating a CF update).

## How to deploy

- adapt routes in manifest.yml
- deploy with

```
cf7 push --no-start
cf7 add-network-policy c2ctest --destination-app c2ctest --protocol tcp --port 1-65535
cf7 logs c2ctest > c2ctest.log
cf7 start c2ctest
```

## Test rolling update of app

```
cf7 restart c2ctest --strategy rolling
```

Interesting logs:
```
# start new instance 10.154.64.54
   2020-04-09T10:12:52.69+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:12:52,697 - 10.154.64.54 - main - INFO - CF_INSTANCE_IP=10.0.73.1
   2020-04-09T10:12:52.69+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:12:52,697 - 10.154.64.54 - main - INFO - CF_INSTANCE_INTERNAL_IP=10.154.64.54
   2020-04-09T10:12:52.69+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:12:52,697 - 10.154.64.54 - main - INFO - ext_route=https://c2ctest.cfapps.dev22.aws.bndl.sapcloud.io/ext/from_10.154.64.54
   2020-04-09T10:12:52.69+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:12:52,697 - 10.154.64.54 - main - INFO - int_route=http://c2ctest.apps.internal:8080/int/from_10.154.64.54
   2020-04-09T10:12:52.69+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:12:52,697 - 10.154.64.54 - main - INFO - PORT=8080
   2020-04-09T10:12:52.69+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:12:52,698 - 10.154.64.54 - main - INFO - Outgoing requests before server start
   2020-04-09T10:12:52.80+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:12:52,805 - 10.154.64.53 - main - INFO - GET /ext/from_10.154.64.54/1 HTTP/1.1
   2020-04-09T10:12:52.89+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:12:52,890 - 10.154.64.54 - main - INFO - GET https://c2ctest.cfapps.dev22.aws.bndl.sapcloud.io/ext/from_10.154.64.54/1: status=200, res=10.154.64.53
# int. request from new to old instance works
   2020-04-09T10:12:52.89+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:12:52,893 - 10.154.64.53 - main - INFO - GET /int/from_10.154.64.54/2 HTTP/1.1
   2020-04-09T10:12:52.89+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:12:52,894 - 10.154.64.54 - main - INFO - GET http://c2ctest.apps.internal:8080/int/from_10.154.64.54/2: status=200, res=10.154.64.53
   2020-04-09T10:12:52.89+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:12:52,895 - 10.154.64.54 - main - INFO - Server start in 10s
   2020-04-09T10:12:56.73+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:12:56,737 - 10.154.64.53 - main - INFO - GET http://c2ctest.apps.internal:8080/int/from_10.154.64.53/42: status=200, res=10.154.64.53
   2020-04-09T10:13:01.75+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:13:01,751 - 10.154.64.53 - main - INFO - GET /ext/from_10.154.64.53/43 HTTP/1.1
   2020-04-09T10:13:01.75+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:13:01,753 - 10.154.64.53 - main - INFO - GET https://c2ctest.cfapps.dev22.aws.bndl.sapcloud.io/ext/from_10.154.64.53/43: status=200, res=10.154.64.53
   2020-04-09T10:13:01.75+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:13:01,755 - 10.154.64.53 - main - INFO - GET /int/from_10.154.64.53/44 HTTP/1.1
   2020-04-09T10:13:01.75+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:13:01,756 - 10.154.64.53 - main - INFO - GET http://c2ctest.apps.internal:8080/int/from_10.154.64.53/44: status=200, res=10.154.64.53
   2020-04-09T10:13:02.91+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:13:02,912 - 10.154.64.53 - main - INFO - GET /ext/from_10.154.64.54/3 HTTP/1.1
   2020-04-09T10:13:02.91+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:13:02,917 - 10.154.64.54 - main - INFO - GET http://c2ctest.apps.internal:8080/int/from_10.154.64.54/4: status=200, res=10.154.64.53
   2020-04-09T10:13:05.17+0200 [CELL/0] OUT Container became healthy
   2020-04-09T10:13:06.76+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:13:06,769 - 10.154.64.53 - main - INFO - GET /ext/from_10.154.64.53/45 HTTP/1.1
   2020-04-09T10:13:06.77+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:13:06,770 - 10.154.64.53 - main - INFO - GET https://c2ctest.cfapps.dev22.aws.bndl.sapcloud.io/ext/from_10.154.64.53/45: status=200, res=10.154.64.53
   2020-04-09T10:13:06.77+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:13:06,773 - 10.154.64.53 - main - INFO - GET /int/from_10.154.64.53/46 HTTP/1.1
   2020-04-09T10:13:06.77+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:13:06,773 - 10.154.64.53 - main - INFO - GET http://c2ctest.apps.internal:8080/int/from_10.154.64.53/46: status=200, res=10.154.64.53
   2020-04-09T10:13:07.93+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:13:07,930 - 10.154.64.53 - main - INFO - GET /ext/from_10.154.64.54/5 HTTP/1.1
   2020-04-09T10:13:07.93+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:13:07,932 - 10.154.64.54 - main - INFO - GET https://c2ctest.cfapps.dev22.aws.bndl.sapcloud.io/ext/from_10.154.64.54/5: status=200, res=10.154.64.53
   2020-04-09T10:13:07.93+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:13:07,935 - 10.154.64.54 - main - INFO - GET /int/from_10.154.64.54/6 HTTP/1.1
   2020-04-09T10:13:07.93+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:13:07,936 - 10.154.64.54 - main - INFO - GET http://c2ctest.apps.internal:8080/int/from_10.154.64.54/6: status=200, res=10.154.64.54
   2020-04-09T10:13:08.05+0200 [CELL/0] OUT Cell 21ed4c93-de7a-4ed1-a5e8-42b6220665ef stopping instance c29d2ba6-edfd-4168-65ba-d5d0
   2020-04-09T10:13:08.31+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:13:08,315 - 10.154.64.53 - main - INFO - Shutting down
   2020-04-09T10:13:08.31+0200 [CELL/SSHD/0] OUT Exit status 0
   2020-04-09T10:13:11.77+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:13:11,777 - 10.154.64.53 - main - INFO - Last outgoing requests before shutdown
   2020-04-09T10:13:11.78+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:13:11,786 - 10.154.64.54 - main - INFO - GET /ext/from_10.154.64.53/47 HTTP/1.1
   2020-04-09T10:13:11.78+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:13:11,787 - 10.154.64.53 - main - INFO - GET https://c2ctest.cfapps.dev22.aws.bndl.sapcloud.io/ext/from_10.154.64.53/47: status=200, res=10.154.64.54
# int. request from old to new instance works until shutdown
   2020-04-09T10:13:11.79+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:13:11,791 - 10.154.64.54 - main - INFO - GET /int/from_10.154.64.53/48 HTTP/1.1
   2020-04-09T10:13:11.79+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:13:11,792 - 10.154.64.53 - main - INFO - GET http://c2ctest.apps.internal:8080/int/from_10.154.64.53/48: status=200, res=10.154.64.54
   2020-04-09T10:13:11.93+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:13:11,936 - 10.154.64.53 - main - INFO - Server stopped
   2020-04-09T10:13:11.99+0200 [APP/PROC/WEB/0] OUT Exit status 0
   2020-04-09T10:13:11.99+0200 [CELL/0] OUT Cell 21ed4c93-de7a-4ed1-a5e8-42b6220665ef destroying container for instance c29d2ba6-edfd-4168-65ba-d5d0
   2020-04-09T10:13:12.57+0200 [CELL/0] OUT Cell 21ed4c93-de7a-4ed1-a5e8-42b6220665ef successfully destroyed container for instance c29d2ba6-edfd-4168-65ba-d5d0
   2020-04-09T10:13:12.94+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:13:12,946 - 10.154.64.54 - main - INFO - GET /ext/from_10.154.64.54/7 HTTP/1.1
   2020-04-09T10:13:12.94+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:13:12,948 - 10.154.64.54 - main - INFO - GET https://c2ctest.cfapps.dev22.aws.bndl.sapcloud.io/ext/from_10.154.64.54/7: status=200, res=10.154.64.54
   2020-04-09T10:13:12.95+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:13:12,951 - 10.154.64.54 - main - INFO - GET /int/from_10.154.64.54/8 HTTP/1.1
   2020-04-09T10:13:12.95+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:13:12,952 - 10.154.64.54 - main - INFO - GET http://c2ctest.apps.internal:8080/int/from_10.154.64.54/8: status=200, res=10.154.64.54
   2020-04-09T10:13:17.96+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:13:17,963 - 10.154.64.54 - main - INFO - GET /ext/from_10.154.64.54/9 HTTP/1.1
   2020-04-09T10:13:17.96+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:13:17,964 - 10.154.64.54 - main - INFO - GET https://c2ctest.cfapps.dev22.aws.bndl.sapcloud.io/ext/from_10.154.64.54/9: status=200, res=10.154.64.54
   2020-04-09T10:13:17.96+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:13:17,966 - 10.154.64.54 - main - INFO - GET /int/from_10.154.64.54/10 HTTP/1.1
   2020-04-09T10:13:17.96+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:13:17,967 - 10.154.64.54 - main - INFO - GET http://c2ctest.apps.internal:8080/int/from_10.154.64.54/10: status=200, res=10.154.64.54
   2020-04-09T10:13:22.99+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:13:22,991 - 10.154.64.54 - main - INFO - GET /ext/from_10.154.64.54/11 HTTP/1.1
   2020-04-09T10:13:22.99+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:13:22,998 - 10.154.64.54 - main - INFO - GET http://c2ctest.apps.internal:8080/int/from_10.154.64.54/12: status=200, res=10.154.64.54
   ...
```

- no errors at all, container networking works for old and new instance without problems

## Test CF update

- find the diego cell on which app instance is running
  - via CF_INSTANCE_IP log on app start, or
  - via `cf7 curl /v2/apps/$(cf7 app c2ctest --guid)/stats`
- recreate diego cell to simulate a CF update

```
bosh -d cf vms | grep IP
bosh -d cf create diego-cell/GUID
```

Interesting logs:
```
# cell draining started by bosh recreate diego-cell, ~08:14 UTC
# start new instance 10.141.66.136
   2020-04-09T10:14:52.09+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:14:52,097 - 10.141.66.136 - main - INFO - CF_INSTANCE_IP=10.0.73.3
   2020-04-09T10:14:52.09+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:14:52,097 - 10.141.66.136 - main - INFO - CF_INSTANCE_INTERNAL_IP=10.141.66.136
   2020-04-09T10:14:52.09+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:14:52,098 - 10.141.66.136 - main - INFO - ext_route=https://c2ctest.cfapps.dev22.aws.bndl.sapcloud.io/ext/from_10.141.66.136
   2020-04-09T10:14:52.09+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:14:52,098 - 10.141.66.136 - main - INFO - int_route=http://c2ctest.apps.internal:8080/int/from_10.141.66.136
   2020-04-09T10:14:52.09+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:14:52,098 - 10.141.66.136 - main - INFO - PORT=8080
# ext. request to old instance 10.154.64.54 works
   2020-04-09T10:14:52.21+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:14:52,219 - 10.154.64.54 - main - INFO - GET /ext/from_10.141.66.136/1 HTTP/1.1
   2020-04-09T10:14:52.29+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:14:52,296 - 10.141.66.136 - main - INFO - GET https://c2ctest.cfapps.dev22.aws.bndl.sapcloud.io/ext/from_10.141.66.136/1: status=200, res=10.154.64.54
# int. request from new instance (probably to old instance) failed
   2020-04-09T10:14:53.30+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:14:53,299 - 10.141.66.136 - main - ERROR - GET http://c2ctest.apps.internal:8080/int/from_10.141.66.136/2: <urlopen error timed out>
   2020-04-09T10:14:53.30+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:14:53,300 - 10.141.66.136 - main - INFO - Server start in 10s
   2020-04-09T10:14:53.37+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:14:53,371 - 10.154.64.54 - main - INFO - GET /ext/from_10.154.64.54/47 HTTP/1.1
   2020-04-09T10:14:53.37+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:14:53,374 - 10.154.64.54 - main - INFO - GET https://c2ctest.cfapps.dev22.aws.bndl.sapcloud.io/ext/from_10.154.64.54/47: status=200, res=10.154.64.54
# int. request from old instance to old instance still works
   2020-04-09T10:14:53.37+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:14:53,376 - 10.154.64.54 - main - INFO - GET /int/from_10.154.64.54/48 HTTP/1.1
   2020-04-09T10:14:53.37+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:14:53,377 - 10.154.64.54 - main - INFO - GET http://c2ctest.apps.internal:8080/int/from_10.154.64.54/48: status=200, res=10.154.64.54
   2020-04-09T10:14:58.39+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:14:58,390 - 10.154.64.54 - main - INFO - GET /ext/from_10.154.64.54/49 HTTP/1.1
   2020-04-09T10:14:58.39+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:14:58,393 - 10.154.64.54 - main - INFO - GET https://c2ctest.cfapps.dev22.aws.bndl.sapcloud.io/ext/from_10.154.64.54/49: status=200, res=10.154.64.54
   2020-04-09T10:14:58.39+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:14:58,395 - 10.154.64.54 - main - INFO - GET /int/from_10.154.64.54/50 HTTP/1.1
   2020-04-09T10:14:58.39+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:14:58,396 - 10.154.64.54 - main - INFO - GET http://c2ctest.apps.internal:8080/int/from_10.154.64.54/50: status=200, res=10.154.64.54
   2020-04-09T10:15:03.31+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:03,310 - 10.141.66.136 - main - INFO - Server started
   2020-04-09T10:15:03.31+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:03,319 - 10.154.64.54 - main - INFO - GET /ext/from_10.141.66.136/3 HTTP/1.1
   2020-04-09T10:15:03.32+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:03,321 - 10.141.66.136 - main - INFO - GET https://c2ctest.cfapps.dev22.aws.bndl.sapcloud.io/ext/from_10.141.66.136/3: status=200, res=10.154.64.54
   2020-04-09T10:15:03.41+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:03,411 - 10.154.64.54 - main - INFO - GET /ext/from_10.154.64.54/51 HTTP/1.1
   2020-04-09T10:15:03.41+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:03,413 - 10.154.64.54 - main - INFO - GET https://c2ctest.cfapps.dev22.aws.bndl.sapcloud.io/ext/from_10.154.64.54/51: status=200, res=10.154.64.54
   2020-04-09T10:15:03.41+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:03,415 - 10.154.64.54 - main - INFO - GET /int/from_10.154.64.54/52 HTTP/1.1
   2020-04-09T10:15:03.41+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:03,416 - 10.154.64.54 - main - INFO - GET http://c2ctest.apps.internal:8080/int/from_10.154.64.54/52: status=200, res=10.154.64.54
# int. request from new instance still doesn't work
   2020-04-09T10:15:04.32+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:04,324 - 10.141.66.136 - main - ERROR - GET http://c2ctest.apps.internal:8080/int/from_10.141.66.136/4: <urlopen error timed out>
   2020-04-09T10:15:05.00+0200 [CELL/0] OUT Container became healthy
# first ext. request handled by new instance
   2020-04-09T10:15:08.43+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:08,435 - 10.141.66.136 - main - INFO - GET /ext/from_10.154.64.54/53 HTTP/1.1
   2020-04-09T10:15:08.43+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:08,437 - 10.154.64.54 - main - INFO - GET https://c2ctest.cfapps.dev22.aws.bndl.sapcloud.io/ext/from_10.154.64.54/53: status=200, res=10.141.66.136
   2020-04-09T10:15:09.34+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:09,340 - 10.141.66.136 - main - INFO - GET https://c2ctest.cfapps.dev22.aws.bndl.sapcloud.io/ext/from_10.141.66.136/5: status=200, res=10.141.66.136
# int. request of new instance to itself: worked
   2020-04-09T10:15:09.34+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:09,343 - 10.141.66.136 - main - INFO - GET /int/from_10.141.66.136/6 HTTP/1.1
   2020-04-09T10:15:09.34+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:09,343 - 10.141.66.136 - main - INFO - GET http://c2ctest.apps.internal:8080/int/from_10.141.66.136/6: status=200, res=10.141.66.136
# but no int. request between the instances observed
   2020-04-09T10:15:09.44+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:09,440 - 10.154.64.54 - main - ERROR - GET http://c2ctest.apps.internal:8080/int/from_10.154.64.54/54: <urlopen error timed out>
   2020-04-09T10:15:11.44+0200 [CELL/0] OUT Cell 21ed4c93-de7a-4ed1-a5e8-42b6220665ef stopping instance e1bfa4d3-aadd-461e-420b-164b
   2020-04-09T10:15:11.50+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:11,507 - 10.154.64.54 - main - INFO - Shutting down
   2020-04-09T10:15:11.50+0200 [CELL/SSHD/0] OUT Exit status 0
   2020-04-09T10:15:14.35+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:14,356 - 10.141.66.136 - main - INFO - GET /ext/from_10.141.66.136/7 HTTP/1.1
   2020-04-09T10:15:14.35+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:14,358 - 10.141.66.136 - main - INFO - GET https://c2ctest.cfapps.dev22.aws.bndl.sapcloud.io/ext/from_10.141.66.136/7: status=200, res=10.141.66.136
   2020-04-09T10:15:14.36+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:14,361 - 10.141.66.136 - main - INFO - GET /int/from_10.141.66.136/8 HTTP/1.1
   2020-04-09T10:15:14.36+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:14,362 - 10.141.66.136 - main - INFO - GET http://c2ctest.apps.internal:8080/int/from_10.141.66.136/8: status=200, res=10.141.66.136
   2020-04-09T10:15:14.44+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:14,443 - 10.154.64.54 - main - INFO - Last outgoing requests before shutdown
   2020-04-09T10:15:14.45+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:14,456 - 10.141.66.136 - main - INFO - GET /ext/from_10.154.64.54/55 HTTP/1.1
   2020-04-09T10:15:14.45+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:14,459 - 10.154.64.54 - main - INFO - GET https://c2ctest.cfapps.dev22.aws.bndl.sapcloud.io/ext/from_10.154.64.54/55: status=200, res=10.141.66.136
# last int. request failure
   2020-04-09T10:15:15.46+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:15,462 - 10.154.64.54 - main - ERROR - GET http://c2ctest.apps.internal:8080/int/from_10.154.64.54/56: <urlopen error timed out>
   2020-04-09T10:15:15.66+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:15,661 - 10.154.64.54 - main - INFO - Server stopped
# old instance is gone, no more int. request errors from here
   2020-04-09T10:15:15.69+0200 [APP/PROC/WEB/0] OUT Exit status 0
   2020-04-09T10:15:15.69+0200 [CELL/0] OUT Cell 21ed4c93-de7a-4ed1-a5e8-42b6220665ef destroying container for instance e1bfa4d3-aadd-461e-420b-164b
   2020-04-09T10:15:16.28+0200 [CELL/0] OUT Cell 21ed4c93-de7a-4ed1-a5e8-42b6220665ef successfully destroyed container for instance e1bfa4d3-aadd-461e-420b-164b
   2020-04-09T10:15:19.42+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:19,424 - 10.141.66.136 - main - INFO - GET /ext/from_10.141.66.136/9 HTTP/1.1
   2020-04-09T10:15:19.42+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:19,425 - 10.141.66.136 - main - INFO - GET https://c2ctest.cfapps.dev22.aws.bndl.sapcloud.io/ext/from_10.141.66.136/9: status=200, res=10.141.66.136
   2020-04-09T10:15:19.42+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:19,428 - 10.141.66.136 - main - INFO - GET /int/from_10.141.66.136/10 HTTP/1.1
   2020-04-09T10:15:19.42+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:19,429 - 10.141.66.136 - main - INFO - GET http://c2ctest.apps.internal:8080/int/from_10.141.66.136/10: status=200, res=10.141.66.136
   2020-04-09T10:15:24.44+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:24,442 - 10.141.66.136 - main - INFO - GET /ext/from_10.141.66.136/11 HTTP/1.1
   2020-04-09T10:15:24.44+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:24,444 - 10.141.66.136 - main - INFO - GET https://c2ctest.cfapps.dev22.aws.bndl.sapcloud.io/ext/from_10.141.66.136/11: status=200, res=10.141.66.136
   2020-04-09T10:15:24.44+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:24,446 - 10.141.66.136 - main - INFO - GET /int/from_10.141.66.136/12 HTTP/1.1
   2020-04-09T10:15:24.44+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:24,447 - 10.141.66.136 - main - INFO - GET http://c2ctest.apps.internal:8080/int/from_10.141.66.136/12: status=200, res=10.141.66.136
   2020-04-09T10:15:29.48+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:29,487 - 10.141.66.136 - main - INFO - GET /ext/from_10.141.66.136/13 HTTP/1.1
   2020-04-09T10:15:29.48+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:29,489 - 10.141.66.136 - main - INFO - GET https://c2ctest.cfapps.dev22.aws.bndl.sapcloud.io/ext/from_10.141.66.136/13: status=200, res=10.141.66.136
   2020-04-09T10:15:29.49+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:29,491 - 10.141.66.136 - main - INFO - GET /int/from_10.141.66.136/14 HTTP/1.1
   2020-04-09T10:15:29.49+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:29,492 - 10.141.66.136 - main - INFO - GET http://c2ctest.apps.internal:8080/int/from_10.141.66.136/14: status=200, res=10.141.66.136
   2020-04-09T10:15:34.50+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:34,505 - 10.141.66.136 - main - INFO - GET /ext/from_10.141.66.136/15 HTTP/1.1
   2020-04-09T10:15:34.50+0200 [APP/PROC/WEB/0] OUT 2020-04-09 08:15:34,507 - 10.141.66.136 - main - INFO - GET https://c2ctest.cfapps.dev22.aws.bndl.sapcloud.io/ext/from_10.141.66.136/15: status=200, res=10.141.66.136
...
```

- errors on container networking observed
- no successful internal request between old and new instance at all
- no errors for external requests

The complete logs of this container networking test is contained in `c2ctest.log`. Bosh recreate logs are appended.

## Possible Improvements

- discover internal addresses via DNS and external requests
- explicitely call all discovered internal IPs to enforce communication between old and new instances

**Comment by KauzClay** [comment]
Hey @stephanme ,

We did some digging, were able to reproduce the issue, and we think we know what is going on.

The drain script for silk-daemon pings the rep to find out if it is still up, and is supposed to wait until rep is finished and goes down before starting (i.e. as long as the ping succeeds, silk-daemon drain will wait). This happens here: https://github.com/cloudfoundry/silk-release/blob/6b08f84a78a4b35605f13b5f79ee7809406c0362/jobs/silk-daemon/templates/drain.erb#L27

However, at some point diego has [changed this to use https](https://github.com/cloudfoundry/diego-release/commit/14a6a5dbe3530b69ba897c40c9a731d250b56196#diff-bb6cc2345a275c0e0efe6a1961543f7b). So, the ping will fail because it is trying to talk http to an https server, and silk-daemon drain will interpret that as the rep being down, so the teardown happens _before_ the rep has finished evacuating, giving you these timeout errors. This is WRONG!

We should make our silk-daemon drain script use https so that we get the expected behavior of waiting for rep to go down before starting to drain.

We will work on fixing this, thanks for catching it!

cc @christianang @mcwumbly

**Comment by KauzClay** [comment]
@cloudfoundry/cf-diego  we think it is pretty likely this issue was caused by this change: https://github.com/cloudfoundry/diego-release/commit/14a6a5dbe3530b69ba897c40c9a731d250b56196#diff-bb6cc2345a275c0e0efe6a1961543f7b

Not sure how you would have known, but silk uses the rep's ping endpoint to determine when it is safe to begin draining. Do you think there's anything we could do in either silk or diego to make this contract more explicit? Maybe even a comment in the implementation of that endpoint would be enough. Just something to highlight that there is some coordination going on here.

**Comment by KauzClay** [comment]
Hey @stephanme ,

We made a fix for this issue on our develop branch. It still needs to make it through our pipelines, but it will be in the next release we cut, silk-release/cf-networking-release 2.30.0.

Thanks!
@KauzClay @christianang

**Comment by stephanme** [comment]
Thanks for taking care 👍

**Comment by XanderStrike** [comment]
The fix is in! Check out release [2.30.0](https://github.com/cloudfoundry/cf-networking-release/releases/tag/2.30.0)
